File: meetingRoomsII.cpp
//time complexity:O(nlogn)
//space complexity:O(n)
//executed on leetcode: yes
//approach:using min heap
//any issues faced? no

class Solution {
public:
    int minMeetingRooms(vector<vector<int>>& intervals) {
        sort(intervals.begin(), intervals.end());
        priority_queue<int, vector<int>, greater<int>>rooms;
        for(int i=0; i<intervals.size(); i++)
        {
            if(rooms.empty())
            {
                rooms.push(intervals[i][1]);
            }
            else if(intervals[i][0]<rooms.top())
            {
                rooms.push(intervals[i][1]);
            }
            else
            {
                rooms.pop();
                rooms.push(intervals[i][1]);
            }
        }
        return rooms.size();
    }
};

File: meetingroomsii.java
// Time Complexity : O(nlogn)
// Space Complexity :O(k)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
    public int minMeetingRooms(int[][] intervals) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        Arrays.sort(intervals, (a,b)->a[0]-b[0]);
        
        for(int[] interval:intervals){
            if(!pq.isEmpty() && pq.peek()<=interval[0]){
                pq.poll();
            }
            pq.add(interval[1]);
        }
        return pq.size();
    }
}